File: Java version of Life Game/Life_Game/Life_Game/Game_field.java
package Main.Life_Game;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;
import java.util.Scanner;

public class Game_field {
    // basic declaration;
    private static int x,y,c,h, empty_tiles_percent;
    private static ArrayList<Herbivore> herbivores;
    private static ArrayList<Carnivore> carnivores;
            // just static for better mobility and comfortability of coding (a bit against oop but irrelevant)
            static String[][] game_field;
            // general random to avoid exception with different randoms
            static Random random = new Random();
    private static Scanner sc = new Scanner(System.in);

    public Game_field(int[] preset){
        x =preset[0];
        y =preset[1];
        c =preset[2];
        h =preset[3];
        empty_tiles_percent =preset[4];
        herbivores=new ArrayList<>();
        carnivores=new ArrayList<>();
        generate_field(x,y, empty_tiles_percent);
    }

    public Game_field(ArrayList<ArrayList<int[]>> preset){
        int[] main_settings  =  preset.get(0).get(0);
        x =main_settings[0];
        y =main_settings[1];
        c =main_settings[2];
        h =main_settings[3];
        empty_tiles_percent =main_settings[4];
        herbivores=new ArrayList<>();
        carnivores=new ArrayList<>();

        Carnivore.setAgeCap(main_settings[5]);
        Herbivore.setAgeCap(main_settings[6]);

        ArrayList<int[]> positions = preset.get(1);
        if (preset.get(1).size()==0){
            generate_field(x,y, empty_tiles_percent);
        }
        else{
            for (int i = 0; i < c; i++) { //carnivores positions
                carnivores.add(new Carnivore(i,positions.get(i),random));
            }
            for (int i = c; i < positions.size(); i++) { //herbivores positions
                herbivores.add(new Herbivore(i,positions.get(i),random));
            }
            generate_field(x,y, empty_tiles_percent,carnivores,herbivores);
        }

    }

    public static void generate_field(int x, int y, int empty_tiles_percent) {
        game_field = new String[y][x];

        for (int i = 0; i < y; i++) {
            Arrays.fill(game_field[i], "-");
        }
        // Calculate the number of grass tiles based on empty_tiles_percent
        int totalTiles = x * y;
        int emptyTiles = (empty_tiles_percent * totalTiles) / 100;
        int grassTiles = totalTiles - emptyTiles;
        // Fill the required number of grass tiles
        while (grassTiles > 0) {
            int row = random.nextInt(y);
            int col = random.nextInt(x);
            if (game_field[row][col].equals("-")) {
                game_field[row][col] = "+";
                grassTiles--;
            }
        }

        // Random emplacement of Carnivores and Herbivores
        int index_h = 0, index_c = 0;
        int c_local=c;
        int h_local=h;
        while (c_local > 0 || h_local > 0) {
            if (c_local > 0) {
                int[] position = {random.nextInt(0, y), random.nextInt(0, x)};
                while (game_field[position[0]][position[1]].equals("H")){
                    position = new int[]{random.nextInt(0, y), random.nextInt(0, x)};
                }
                //carnivores[index_c] = new Carnivore(index_c,position, random);
                carnivores.add(new Carnivore(index_c,position, random));
                index_c++;
                game_field[position[0]][position[1]] = "C";
                c_local--;
            }
            if (h_local > 0) {
                int[] position = {random.nextInt(0, y), random.nextInt(0, x)};
                while (game_field[position[0]][position[1]].equals("C")){
                    position = new int[]{random.nextInt(0, y), random.nextInt(0, x)};
                }
                //herbivores[index_h] = new Herbivore(index_h,position, random);
                herbivores.add(new Herbivore(index_h,position, random));
                index_h++;
                game_field[position[0]][position[1]] = "H";
                h_local--;
            }
        }
    }
    public static void generate_field(int x, int y, int empty_tiles_percent, ArrayList<Carnivore> carnivores, ArrayList<Herbivore> herbivores) {
        game_field = new String[y][x];

        for (int i = 0; i < y; i++) {
            Arrays.fill(game_field[i], "-");
        }
        // Calculate the number of grass tiles based on empty_tiles_percent
        int totalTiles = x * y;
        int emptyTiles = (empty_tiles_percent * totalTiles) / 100;
        int grassTiles = totalTiles - emptyTiles;
        // Fill the required number of grass tiles
        while (grassTiles > 0) {
            int row = random.nextInt(y);
            int col = random.nextInt(x);
            if (game_field[row][col].equals("-")) {
                game_field[row][col] = "+";
                grassTiles--;
            }
        }

        // Emplacement of Carnivores and Herbivores
        for (Carnivore cv: carnivores){
            int[] l_position = cv.getPosition();
            game_field[l_position[0]][l_position[1]]="C";
        }
        for(Herbivore hb: herbivores){
            int[] l_position = hb.getPosition();
            game_field[l_position[0]][l_position[1]]="H";
        }

    }


    public static void play(){
        Actions.eat_everybody();
        Actions.refresh();
        Actions.move_everybody();
        Actions.refresh();
        print();
    }


    public static void print(){
        int c_g=0;
        int c_e=0;
        for (int i = 0; i < game_field.length; i++) {
            for (int j = 0; j < game_field[i].length; j++) {
                System.out.print(game_field[i][j]+" ");
                if(game_field[i][j].equals("-")){
                    c_e++;
                }else if (game_field[i][j].equals("+")){
                    c_g++;
                }
            }
            System.out.println();
        }

        System.out.println("CARNIVORES LEFT: "+getCarnivores().size());
        System.out.println("HERBIVORES LEFT: "+getHerbivores().size());
        System.out.println("GRASS: "+c_g);
        System.out.println("EMPTY: "+c_e);
        System.out.println("MOVE ITERATION: "+Actions.getIteration());
        System.out.println("CARNIVORES DEAD: "+ (getC()-getCarnivores().size()));
        System.out.println("HERBIVORES DEAD: "+ (getH()-getHerbivores().size()));

        if(Actions.isGame_over_E()){
            System.out.println("The game is over. Everybody is dead!");
            System.exit(0);
        }
        else if(Actions.isGame_over_C()){
            System.out.println("The game is over. All carnivores are dead!");
            System.exit(0);
        }
        else if(Actions.isGame_over_H()){
            System.out.println("The game is over. All herbivores are dead!");
            System.exit(0);
        }

            // looping the game process
            System.out.println("TO CONTINUE, ENTER 1; ELSE 0");
            String response = sc.next();
            if(response.equals("1")){
                play();
                // forced game shutdown
            }else if (response.equals("0")){
                System.exit(0);
            }


    }

    // GETTERS AND SETTERS

    public static int getX() {
        return x;
    }

    public static int getY() {
        return y;
    }

    public static int getH() {
        return h;
    }

    public static int getC() {
        return c;
    }

    public static ArrayList<Herbivore> getHerbivores() {
        return herbivores;
    }

    public static ArrayList<Carnivore>  getCarnivores() {
        return carnivores;
    }

    public static Random getRandom() {
        return random;
    }

    public static String getTile(int[] position){
        return Game_field.game_field[position[0]][position[1]];
    }

    public static Herbivore getHerbivoreByPosition(int[] position){
        //Herbivore[] herbivores = getHerbivores();
        ArrayList<Herbivore> herbivores = getHerbivores();
        for (Herbivore herbivore : herbivores) {
            if (Arrays.equals(herbivore.getPosition(), position))
                return herbivore;
        }
        return null;
    }
    public static Carnivore getCarnivoreByPosition(int[] position){
        //Carnivore[] carnivores = getCarnivores();
        ArrayList<Carnivore> carnivores = getCarnivores();
        for (Carnivore carnivore : carnivores) {
            if (Arrays.equals(carnivore.getPosition(), position))
                return carnivore;
        }
        return null;
    }

}

File: newVersion/Game_field.cpp
#include "Game_field.h"
#include "Herbivore.h"
#include "Carnivore.h"
#include "Actions.h"
#include "Carnivore.cpp"

//static members have to be declared beforehand
int Game_field::x;
int Game_field::y;
int Game_field::c;
int Game_field::h;
int Game_field::empty_tiles_percent;
vector<Carnivore> Game_field::carnivores;
vector<Herbivore> Game_field::herbivores;
vector<vector<string>> Game_field::game_field;

//constructor for basic preset
Game_field::Game_field(vector<int> preset){
    x=preset[0];
    y=preset[1];
    c=preset[2];
    h=preset[3];
    empty_tiles_percent=preset[4];

    carnivores.reserve(c);
    herbivores.reserve(h);
    
    generateField(x,y,empty_tiles_percent);
}

// constructor for fully customized preset
Game_field::Game_field(vector<vector<vector<int>>> preset){
    vector<int> mainSettings = preset[0][0];
    x = mainSettings[0];
    y = mainSettings[1];
    c = mainSettings[2];
    h = mainSettings[3];
    empty_tiles_percent = mainSettings[4];
        
    herbivores.reserve(h);
    carnivores.reserve(c);
        
    Carnivore::setAgeCap(mainSettings[5]);
    Herbivore::setAgeCap(mainSettings[6]);

    vector<vector<int>> positions = preset[1];

    if (positions.empty()) {
         generateField(x,y,empty_tiles_percent);
    } else {
        for (int i = 0; i < c; ++i) {
            carnivores.emplace_back(i, positions[i]);
        }

        for (int i = c; i < positions.size(); ++i) {
            herbivores.emplace_back(i, positions[i]);
        }

        generateField(x,y,empty_tiles_percent,carnivores, herbivores);
    }
}

// simplistic creation of gamefield 
void Game_field::generateField(int x, int y, int empty_tiles_percent){
        game_field = vector<vector<string>>(y, vector<string>(x, "-"));
        // Calculate the number of grass tiles based on empty_tiles_percent
        int totalTiles = x * y;
        int emptyTiles = (empty_tiles_percent * totalTiles) / 100;
        int grassTiles = totalTiles - emptyTiles;
        // Fill the required number of grass tiles
        while (grassTiles > 0) {
            int row = rand() % y;
            int col = rand() % x;

            if (game_field[row][col] == "-") {
                game_field[row][col] = "+";
                grassTiles--;
            }
        }
        // Random emplacement of Carnivores and Herbivores
        int indexH = 0, indexC = 0;
        int localC = getC();
        int localH = getH();

        while (localC > 0 || localH > 0) {
            if (localC > 0) {
                int row = rand() % y;
                int col = rand() % x;

                while (game_field[row][col] == "H") {
                    row = rand() % y;
                    col = rand() % x;
                }

                //carnivores.emplace_back(indexC++, array<int, 2>{row, col}, "C");
                Carnivore cv(indexC,{row,col});
                carnivores.emplace_back(cv);
                game_field[row][col] = "C";
                localC--;
            }

            if (localH > 0) {
                int row = rand() % y;
                int col = rand() % x;

                while (game_field[row][col] == "C") {
                    row = rand() % y;
                    col = rand() % x;
                }

                //herbivores.emplace_back(indexH++, array<int, 2>{row, col}, "H");
                Herbivore hb(indexC,{row,col});
                herbivores.emplace_back(hb);
                game_field[row][col] = "H";
                localH--;
            }
        }
}

//intricate creation of gamefield
void Game_field::generateField(int x, int y, int empty_tiles_percent, vector<Carnivore> carnivores, vector<Herbivore> herbivores){
    Game_field::game_field = vector<vector<string>>(y, vector<string>(x, "-"));
        // Calculate the number of grass tiles based on empty_tiles_percent
        int totalTiles = x * y;
        int emptyTiles = (empty_tiles_percent * totalTiles) / 100;
        int grassTiles = totalTiles - emptyTiles;
        // Fill the required number of grass tiles
        while (grassTiles > 0) {
            int row = rand() % y;
            int col = rand() % x;

            if ( Game_field::game_field[row][col] == "-") {
                Game_field::game_field[row][col] = "+";
                grassTiles--;
            }
        }
        // Emplacement of Carnivores and Herbivores
        for (Carnivore cv : carnivores) {
            vector<int> position = cv.getPosition();
            Game_field::game_field[position[0]][position[1]] = "C";
        }

        for (Herbivore hb : herbivores) {
            vector<int> position = hb.getPosition();
            Game_field::game_field[position[0]][position[1]] = "H";
        }
}

void Game_field::print(){
    int c_g=0;
    int c_e=0;

    for(int i =0;i<game_field.size();i++){
        for (int j = 0; j < game_field[i].size(); j++){
            cout<<game_field[i][j]<<" ";
            if(game_field[i][j].compare("-")==0){
                c_e++;
            }else if(game_field[i][j].compare("+")==0) {
                c_g++;
            }
        }
        cout<<endl;
    }
        
    cout<<"Carnivores left: "<<getCarnivores().size()<<endl;
    cout<<"Herbivores left: "<<getHerbivores().size()<<endl;
    cout<<"Grass tiles left: "<<c_g<<endl;
    cout<<"Empty tiles left: "<<c_e<<endl;
    cout<<"MOVE ITERATION: "<<Actions::getIteration();
    cout<<"Carnivores dead: "<<getC()-getCarnivores().size()<<endl;
    cout<<"Herbivores dead: "<<getH()-getHerbivores().size()<<endl;
        
    // if gameover....  
    if(Actions::isGameOverE()){
            cout<<"The game is over. Everybody is dead!\n";
            exit(0);
    }
    else if(Actions::isGameOverC()){
        cout<<"The game is over. All carnivores are dead!\n";
        exit(0);
    }
    else if(Actions::isGameOverH()){
        cout<<"The game is over. All herbivores are dead!\n";
        exit(0);
    }

    // looping the game process
    cout<<"TO CONTINUE, ENTER 1; ELSE 0\n";
    string response;
    getline(cin,response);
    if(response.compare("1")==0){
        play();
        // forced game shutdown
    }else if (response.compare("0")==0){
        exit(0);
    }
}


int Game_field::getX(){return x;}
int Game_field::getY(){return y;}
int Game_field::getC(){return c;}
int Game_field::getH(){return h;}

vector<Herbivore> Game_field::getHerbivores(){return herbivores;}
vector<Carnivore> Game_field::getCarnivores(){return carnivores;}

    //getRandom?
string Game_field::getTile(vector<int> position){
    return Game_field::game_field[position[0]][position[1]];
    }

Herbivore Game_field::getHerbivoreByPosition(vector<int> position){
    vector<Herbivore> herbivores = getHerbivores();
    for (Herbivore hb : herbivores) {
        if(hb.getPosition()==position){
            return hb;
        }
    }
    //return NULL;;
}
Carnivore Game_field::getCarnivoreByPosition(vector<int> position){
    vector<Carnivore> carnivores = getCarnivores();
    for(Carnivore cv : carnivores){
        if(cv.getPosition()==position){
            return cv;
        }
    //return NULL;;
    }
}